libc/src/__support/UInt.h

```diff
@@ -17,79 +17,293 @@
 #include "src/__support/macros/attributes.h"   // LIBC_INLINE
 #include "src/__support/macros/optimization.h" // LIBC_UNLIKELY
 #include "src/__support/macros/properties/types.h" // LIBC_TYPES_HAS_INT128, LIBC_TYPES_HAS_INT64
-#include "src/__support/math_extras.h" // SumCarry, DiffBorrow
+#include "src/__support/math_extras.h" // add_with_carry, sub_with_borrow
 #include "src/__support/number_pair.h"
 
 #include <stddef.h> // For size_t
 #include <stdint.h>
 
 namespace LIBC_NAMESPACE {
 
-namespace internal {
-template <typename T> struct half_width;
+namespace multiword {
 
-template <> struct half_width<uint64_t> : cpp::type_identity<uint32_t> {};
-template <> struct half_width<uint32_t> : cpp::type_identity<uint16_t> {};
+// A type trait mapping unsigned integers to their half-width unsigned
+// counterparts.
+template <typename T> struct half_width;
 template <> struct half_width<uint16_t> : cpp::type_identity<uint8_t> {};
+template <> struct half_width<uint32_t> : cpp::type_identity<uint16_t> {};
+#ifdef LIBC_TYPES_HAS_INT64
+template <> struct half_width<uint64_t> : cpp::type_identity<uint32_t> {};
 #ifdef LIBC_TYPES_HAS_INT128
 template <> struct half_width<__uint128_t> : cpp::type_identity<uint64_t> {};
 #endif // LIBC_TYPES_HAS_INT128
-
+#endif // LIBC_TYPES_HAS_INT64
 template <typename T> using half_width_t = typename half_width<T>::type;
 
-template <typename T> constexpr NumberPair<T> full_mul(T a, T b) {
-  NumberPair<T> pa = split(a);
-  NumberPair<T> pb = split(b);
-  NumberPair<T> prod;
+// An array of two elements that can be used in multiword operations.
+template <typename T> struct Double final : cpp::array<T, 2> {
```

Comment:
I would request you not use the name `Double` since that's already a floating point type. `NumberPair` might work better?

Reply:
Yes, especially a lot of time when `T` is `double`, the result one is called `double-double` type.

Reply:
This type lives in the `multiword` namespace which hosts implementation details for bigint. It is not to be used outside of this file and -in this context- it cannot refer to floating point types. Double is the classic naming for this scheme, e.g.
 - https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0103r1.html
 - https://open-std.org/Jtc1/sc22/wg21/docs/papers/2017/p0104r1.html

Reply:
Fixed in [3e73cd6](https://github.com/llvm/llvm-project/pull/86137/commits/3e73cd668d39eb2c90e93685a31c62720e99c990)